rust/src/mqtt/parser.rs

```diff
@@ -0,0 +1,631 @@
+                 }
+               )
+       ));
+
+#[inline]
+fn parse_disconnect(
+    input: &[u8],
+    mut remaining_len: usize,
+    protocol_version: u8,
+) -> IResult<&[u8], MQTTDisconnectData> {
+    if protocol_version < 5 {
+        return Ok((
+            input,
+            MQTTDisconnectData {
+                reason_code: None,
+                properties: None,
+            },
+        ));
+    }
+    if remaining_len == 0 {
+        // The Reason Code and Property Length can be omitted if the Reason
+        // Code is 0x00 (Normal disconnection) and there are no Properties.
+        // In this case the DISCONNECT has a Remaining Length of 0.
+        return Ok((
+            input,
+            MQTTDisconnectData {
+                reason_code: Some(0),
+                properties: None,
+            },
+        ));
+    }
+    match be_u8(input) {
+        Ok((rem, reason_code)) => {
+            remaining_len -= 1;
+            if remaining_len == 0 {
+                // no properties
+                return Ok((
+                    rem,
+                    MQTTDisconnectData {
+                        reason_code: Some(0),
+                        properties: None,
+                    },
+                ));
+            }
+            match parse_properties(rem, true) {
+                Ok((rem, properties)) => {
+                    return Ok((
+                        rem,
+                        MQTTDisconnectData {
+                            reason_code: Some(reason_code),
+                            properties: properties,
+                        },
+                    ));
+                }
+                Err(e) => return Err(e),
+            }
+        }
+        Err(e) => return Err(e),
+    }
+}
+
+named!(#[inline], pub parse_auth<MQTTAuthData>,
+       do_parse!(
+           reason_code: be_u8
+           >> properties: call!(parse_properties, true)
+           >>  (
+                 MQTTAuthData {
+                   reason_code: reason_code,
+                   properties: properties,
+                 }
+               )
+       ));
+
+pub fn parse_message(input: &[u8], protocol_version: u8) -> IResult<&[u8], MQTTMessage> {
+    // Parse the fixed header first. This is identical across versions and can
+    // be between 2 and 5 bytes long.
+    match parse_fixed_header(input) {
+        Ok((fullrem, header)) => {
+            let len = header.remaining_length as usize;
+            // This is the length of the fixed header that we need to skip
+            // before returning the remainder. It is the sum of the length
+            // of the flag byte (1) and the length of the message length
+            // varint.
+            let skiplen = 1 + varint_length(len);
+            let message_type = header.message_type;
+            if fullrem.len() < len {
+                return Err(Err::Incomplete(Needed::Size(len - fullrem.len())));
+            }
```

Comment:
Is there a danger of DOS/memory exhaustion with specifying a very big `header.remaining_length` ?

Reply:
I think it's easier to not calculate any specific size any more and just request 1 more byte, as discussed earlier in another thread. It's just not worth risking to miss something when in the majority of cases the missing data is there with the next packet  (and hence parser call) anyway.

Reply:
Also, the remaining length is at most 256MB due to protocol range constraints (https://docs.oasis-open.org/mqtt/mqtt/v3.1.1/csprd02/mqtt-v3.1.1-csprd02.html#_Toc385349759). There's not much one can do about that.

Reply:
> I think it's easier to not calculate any specific size any more and just request 1 more byte, as discussed earlier in another thread

Indeed, but you seem to misunderstand my point.

256 MB looks quite big with respect to DOS by memory exhaustion. The TCP streaming engine will bufferize these 256 Mb. So, if I start XX connections and get into this state for these connections and send data slower as it gets more bufferized, Suricata will run out of memory...

> There's not much one can do about that.

Well, we can either 
- at least set a limit, and set an event when the limit is crossed
- have a more "streaming parser" keeping a more complex state, so as not to bufferize all the data (yes this is hard :-/ maybe this 256 Mb is not reachable by every message type to start)

Is this clearer ?

Reply:
Indeed it is.

I get your point about the potential for DOS. Unfortunately MQTT is optimized for space efficiency and does not provide a lot of information upfront about where the space is used, be it for topics, message payloads, etc.  -- you have to parse it to find out! 
So I would propose to introduce a maximum message size to inspect that can be configured in the `mqtt` YAML config. Messages with an indicated length exceeding that would be logged as events, but not parsed until they are complete. What do you think? I think such an approach would at least make the problem manageable. Another idea would be to introduce a memcap value for the parser, and to stop parsing new messages once that memcap has been hit.
Any comments?

Reply:
In this worst case scenario of a 256mb record, what would we do with that data? If nothing much, we can just stream it through. In the smb and nfs parsers I've implemented some logic to track how many bytes to 'skip'. So the data still flows though the parser, but it doesn't do much except the 'skip' bookkeeping. Maybe that could be used here as well?

Reply:
Indeed, SMB has an hardcoded limit of 1Mbyte if I remember correctly.
Configurable limit is even better :-)

Reply:
> In this worst case scenario of a 256mb record, what would we do with that data? If nothing much, we can just stream it through.

Well, the payload is logged in the EVE output -- so if someone sends a big binary payload (like a HD camera image) that would end up in the payload field, escaped for JSON.

> In the smb and nfs parsers I've implemented some logic to track how many bytes to 'skip'. So the data still flows though the parser, but it doesn't do much except the 'skip' bookkeeping. Maybe that could be used here as well?

Well, skipping means not storing data, so there is nothing to output. One could decide to leave out the parsed data in a  large message (i.e. a message that exceeds the limit), set a flag (similar to the filestore TRUNCATED flag) and simply report the message, message type, etc. without the type-specific metadata. That would be a good compromise IMHO - you know that there has been a large message, there are communication and flow metadata, but there is no requirement to parse (and hence buffer) the whole message.

Reply:
> > In this worst case scenario of a 256mb record, what would we do with that data? If nothing much, we can just stream it through.
> 
> Well, the payload is logged in the EVE output -- so if someone submits or receives a big binary payload (like a HD camera image, etc.) that would end up in the payload field, escaped for JSON.

@jasonish we should probably define an upper limit for both the total eve record size as well as individual fields sizes. Can't imagine the mayhem 256mb records would cause :)

Reply:
+1 :)

Reply:
> > > In this worst case scenario of a 256mb record, what would we do with that data? If nothing much, we can just stream it through.
> > 
> > 
> > Well, the payload is logged in the EVE output -- so if someone submits or receives a big binary payload (like a HD camera image, etc.) that would end up in the payload field, escaped for JSON.
> 
> @jasonish we should probably define an upper limit for both the total eve record size as well as individual fields sizes. Can't imagine the mayhem 256mb records would cause :)

2 ideas here. We could enforce an upper bound on the size the builder will build. But I think the parser should be aware, and do truncation if necessary, so you don't lose all context on the record.

Reply:
Currently looking at this -- we can simply decide, after parsing the 2-5 byte fixed header, whether we want to parse the rest of the record or not depending on a configurable maximum parse depth. However, AFAICS we would not be able to partially parse the first portion of the data up to the limit, at least I don't know how to do this in nom without ditching the macros and manually checking the remaining capacity after each parsed item. Parsing a complete record with `do_parse!` would require to read (i.e. buffer!) everything before the final parsed struct is returned with an `IResult` of `Ok`, and that's what we would like to avoid.

My idea was to simply check whether the specified length exceeds the limit and then return and log a simplified special message type that only contains the type and skipped length of the large record. After that I would simply accept (without applayer buffering) and eat upcoming data until we can continue parsing the next message in the connection.
That means we would get something like:
```json
{
  "timestamp": "2020-06-19T16:59:24.964522+0200",
  "flow_id": 1025328194301771,
  "pcap_cnt": 27,
  "event_type": "mqtt",
  "src_ip": "0000:0000:0000:0000:0000:0000:0000:0001",
  "src_port": 50509,
  "dest_ip": "0000:0000:0000:0000:0000:0000:0000:0001",
  "dest_port": 1883,
  "proto": "TCP",
  "mqtt": {
    "publish": {
      "qos": 1,
      "retain": false,
      "dup": false,
      "truncated": true,
      "skipped_length": 1000007
    }
  }
}
```
Such an entry has the message type, the message size and context (src/dest, timestamp, ...) but no additional metadata since we decided not to parse it. In particular, we don't have a message ID to connect it to a transaction (i.e. in the case above we could not connect it to the `puback` that could follow). This is because most of the 256MB could already be used before reaching the position in the record where the message ID occurs.

Would you be OK with these restrictions? I would agree that this is a rare case because I would expect most of the MQTT traffic out there to be smaller than any reasonable limit (i.e. 1MB).

Reply:
Yes, that is what I had in mind :-)

Reply:
Got it :)